### PR TITLE
[GEOT-7163] CQL2-JSON add support for conversion from filters to JSON

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOT-7163](https://osgeo-org.atlassian.net/browse/GEOT-7163) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

working on expression

todate expression

literals

comparisons

and or

geometry contains

basic geometry

time

more functions

cleanup

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

File: modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/CQL2Json.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.geotools.filter.text.cql2.CQLException;
+import org.opengis.filter.Filter;
+
+/**
+ * <b>OGC API CQL2-Json</b>. This class presents the operations available to convert filters into
+ * the CQL2-Json language
+ *
+ * <p>
+ */
+public class CQL2Json {
```
Comment (aaime) [inline modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/CQL2Json.java:31]: I'd add static methods handling Expression as well, mirroring CQL2, CQL and ECQL.

File: modules/unsupported/cql-json/src/test/java/org/geotools/filter/text/cqljson/CQLJsonTest.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package org.geotools.filter.text.cqljson;
```
Comment (aaime) [inline modules/unsupported/cql-json/src/test/java/org/geotools/filter/text/cqljson/CQLJsonTest.java:1]: Copyright header missing  (can you add it to all test classes?)

File: modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/FilterToCQL2Json.java [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+        Expression expr = filter.getExpression();
+        expr.accept(new ExpressionToCQL2Json(objectMapper), args);
+        output.set(ARGS, args);
+
+        return output;
+    }
+
+    @Override
+    public Object visit(BBOX filter, Object extraData) {
+        throw unsupported("BBOX");
```
Comment (aaime) [inline modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/FilterToCQL2Json.java:403]: Ouch, a BBOX filter is very common, it should be supported. Can be translated to an intersection filter if needed?

**Review by aaime — CHANGES_REQUESTED** [review]

File: modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/CQL2Json.java (hunk 1 above)
Comment (turingtestfail) [reply modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/CQL2Json.java:34]: Added

File: modules/unsupported/cql-json/src/test/java/org/geotools/filter/text/cqljson/CQLJsonTest.java (hunk 2 above)
Comment (turingtestfail) [reply modules/unsupported/cql-json/src/test/java/org/geotools/filter/text/cqljson/CQLJsonTest.java:17]: Added

File: modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/FilterToCQL2Json.java (hunk 3 above)
Comment (turingtestfail) [reply modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/FilterToCQL2Json.java]: Added as intersection

Comment (aaime) [reply modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/FilterToCQL2Json.java]: Good. I don't see a test covering this code path, did I miss it?

File: modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/CQL2Json.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        return (JsonNode) filter.accept(toCQL2Json, objectMapper.createObjectNode());
+    }
+
+    /**
+     * Generates the CQL2-Json predicate as a string associated to the {@link Filter} object.
+     *
+     * @param filter the filter to convert
+     * @return the CQL2-Json predicate as a string
+     */
+    public static String toCQL2JsonString(Filter filter) {
```
Comment (aaime) [inline modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/CQL2Json.java:58]:
```suggestion
    public static String toCQL2(Filter filter) {
```

It's clear that it's a JSON target by the class name.

**Review by aaime — CHANGES_REQUESTED** [review]

File: modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/FilterToCQL2Json.java (hunk 3 above)
Comment (turingtestfail) [reply modules/unsupported/cql-json/src/main/java/org/geotools/filter/text/cqljson/FilterToCQL2Json.java]: https://github.com/geotools/geotools/blob/c694004b414dc93c3a631de2e2f4e1a175ea461f/modules/unsupported/cql-json/src/test/java/org/geotools/filter/text/cqljson/FilterToJsonTest.java#L243

**Review by aaime — APPROVED** [review]

**Pull request merged**